lib/url_parser.js

```diff
@@ -4,9 +4,71 @@ var ReadPreference = require('./read_preference'),
   parser = require('url'),
   f = require('util').format,
   assign = require('./utils').assign,
-  Logger = require('mongodb-core').Logger;
+  Logger = require('mongodb-core').Logger,
+  dns = require('dns');
 
-module.exports = function(url, options) {
+module.exports = function(url, options, callback) {
+  var result = parser.parse(url, true);
+  if (result.protocol !== 'mongodb:') {
+    if (result.protocol !== 'mongodb+srv:') {
+      callback(new Error('invalid schema, expected mongodb or mongodb+srv'));
+    }
+  }
+
+  if (result.protocol === 'mongodb+srv:') {
+    if (result.port) {
+      return callback(new Error('Ports not accepted with mongodb+srv'));
+    }
+
+    const srvAddress = `_mongodb._tcp.${result.host}`;
+    dns.resolveSrv(srvAddress, function(err, addresses) {
+      if (err) return callback(err);
+
+      if (addresses.length === 0) {
+        return callback(new Error('No addresses found at host'));
+      }
+
+      dns.resolveTxt(result.host, function(err, records) {
+        if (err && err.code !== 'ENODATA') return callback(err);
+        if (err && err.code === 'ENODATA') records = null;
+
+        let connectionStrings = addresses.map(function(address, i) {
+          if (i === 0) return `mongodb://${address.name}:${address.port}`;
+          else return `${address.name}:${address.port}`;
+        });
+
+        // let connectionStrings = addresses.map(
+        //   (address, i) =>
+        //     i === 0
+        //       ? `mongodb://${address.name}:${address.port}`
+        //       : `${address.name}:${address.port}`
+        // );
+
+        // let connectionString = `mongodb://${addresses
+        //   .map(address => `${address.name}:${address.port}`)
+        //   .join(',')}`;
+
+        let connectionString = connectionStrings.join(',');
+        if (records) connectionString += '/?' + records.join('&');
```

Comment:
should we be doing any sort of sanitization on the query string to prevent bad stuff from happening?